### ISSUE TITLE
Allow "Physical" interfaces to be parents of "Virtual" interfaces

### ISSUE THREAD

**Issue by mngan** [nautobot/nautobot#929]
### Environment
* Python version: 3.6
* Nautobot version: 1.1.3

### Proposed Functionality

Allow the parent relationship extend to arbitrary interfaces instead of just the `lag` type. 

### Use Case

This will allow subinterfaces to have a relationship with the physical interface they are associated with, which will potentially allow me to find a relationship in Graphql between two subinterfaces, at least this is my hope. This will still need some work in a transposer function to tidy up the data, but the alternative is to make a ton of calls to find this information in the transposer. Not sure if the Graphql query could do this any better though. 

Something like this?

```
interfaces(type: "VIRTUAL") {
  parent {
    connected_interface {
      name
      children(type: "VIRTUAL") {
        ...
      }
  }
}
```

### Database Changes

Unsure what will be required, but definitely will need work

### External Dependencies

None that I could anticipate.

### To Do:

- [x] #1455

**Comment by glennmatthews** [comment]
Related: https://github.com/netbox-community/netbox/issues/1519

**Comment by glennmatthews** [comment]
While it's possible to add a custom relationship (Interface -> Interface) to achieve the stated goal of the original issue, it's worth putting some thought into this issue and how we can more generally support inter-interface relationships in core.

See also #1187 and https://github.com/nautobot/nautobot/issues/838#issuecomment-909889212. 

---

By way of comparison, in NetBox 3.1 an Interface or VMInterface has 3 different potential ForeignKeys to another Interface or VMInterface:

- `lag` - reference from a interface (of any type *except* `virtual`) to an interface of type `lag` (note that this is enforced in the form, *not* in the model itself!) on the same device or same virtual-chassis (Interface only, not applicable to VMInterface)
- `parent` - reference from an interface (must be type `virtual`) to an interface of any type on the same device, virtual machine, or virtual-chassis (added to Interface in https://github.com/netbox-community/netbox/pull/5930, extended to VMInterface as well in https://github.com/netbox-community/netbox/commit/a3721a94ce4f13500d98eced74b7d5a4b84817d3, generalized to BaseInterface model in https://github.com/netbox-community/netbox/pull/7622)
- `bridge` - reference from an interface of any type to an interface of any type on the same device, virtual machine, or virtual-chassis (https://github.com/netbox-community/netbox/pull/7622)
  - additionally a specific virtual interface type of `bridge` was added in https://github.com/netbox-community/netbox/commit/61d2158f7633d14f6d6934e0bb0f48c678029691, but `bridge` assignment can involve any types of interfaces, not just this type.

Additionally, the following NetBox feature requests are relevant:

- https://github.com/netbox-community/netbox/issues/5384 - allowing virtual interfaces to be children of a "LAG" such as `bond0` - closed without implementation
- https://github.com/netbox-community/netbox/issues/5402 - allowing a "LAG" such as `bond0` to be a child of a "bridge" such as `vmbr0` - this is covered by https://github.com/netbox-community/netbox/pull/7622 but I wanted to call it out here as specifically allowing a "tree" of subinterface -> interface -> LAG -> bridge interfaces.
- https://github.com/netbox-community/netbox/issues/7552 - adding support for a "breakout" interface type (to e.g. represent a 40G interface broken out into 4x10G interfaces), which would be considered virtual but still assignable to a LAG, and a "parent_position" interface field to indicate interface position within the breakout

---

Offhand it *feels* like NetBox's trio of foreign keys are somewhat redundant, as `lag` and `parent` are currently mutually exclusive fields at least - are there practical use cases where an interface has both a `parent` and a `bridge`, or a `lag` and a `bridge`, that couldn't be (or wouldn't be better) represented by adding the parent or LAG interface to the bridge itself? But perhaps there's some value to having distinct semantics for these three fields even if they end up mostly working the same way under the hood?

**Comment by glennmatthews** [comment]
Separate from the data model considerations, it would be really nice from a UI perspective to, when subinterfaces are present, group the subinterfaces under their parent interfaces with a level of indentation, instead of the current flat listing. Making the sub-groupings collapsable would be even nicer.

**Comment by bryanculver** [comment]
Closed via #1651
